### PR TITLE
Disambiguate 2-n flags in governance new-committee action

### PR DESCRIPTION
**Pull request by smelc**
# Changelog

```yaml
- description: |
    A number of flags appeared in the parameters of new-committee multiple times, with the same long flag, making the parser break. This PR disambiguates the repeated occurences.
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way <- Clément: I believe right?
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Contributes to fixing https://github.com/input-output-hk/cardano-cli/issues/300

# How to trust this PR

This PR is mostly boilerplate, but there is one thing to make sure of:

When a `prefix` argument is introduced in a parser, it is used _in all branches_ of the parser. For example in this snippet:

```hs
asum [ AnyStakePoolKey <$> pStakePoolVerificationKeyOrHashOrFile prefix
       , AnyStakeKey <$> pStakeVerificationKeyOrHashOrFile prefix
       ]
```

It is important to pass `prefix` to every branch of the `asum`. Otherwise we could create inconsistent parsers. I have made sure this was correct.

# Checklist

- [X] Commit sequence broadly makes sense and commits have useful messages
- [X] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- NA The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [X] The changelog section in the PR is updated to describe the change
- [X] Self-reviewed the diff

### CODE REVIEW

**Comment by smelc** [comment]
On the current state of this PR, the `create-new-committee` API is now as follows (cc @CarlosLopezDeLara):

```
Usage: cardano-cli conway governance action create-new-committee 
                                                                   ( --mainnet
                                                                   | --testnet
                                                                   )
                                                                   --governance-action-deposit NATURAL
                                                                   ( --stake-pool-verification-key STRING
                                                                   | --cold-verification-key-file FILE
                                                                   | --stake-pool-id STAKE_POOL_ID
                                                                   | --stake-verification-key STRING
                                                                   | --stake-verification-key-file FILE
                                                                   | --stake-key-hash HASH
                                                                   )
                                                                   --proposal-url TEXT
                                                                   ( --proposal-text TEXT
                                                                   | --proposal-file FILE
                                                                   | --proposal-hash HASH
                                                                   )
                                                                   [ --remove-cc-stake-pool-verification-key STRING
                                                                   | --remove-cc-cold-verification-key-file FILE
                                                                   | --remove-cc-stake-pool-id STAKE_POOL_ID
                                                                   | --remove-cc-stake-verification-key STRING
                                                                   | --remove-cc-stake-verification-key-file FILE
                                                                   | --remove-cc-stake-key-hash HASH
                                                                   ]
                                                                   [
                                                                     ( --add-cc-stake-pool-verification-key STRING
                                                                     | --add-cc-cold-verification-key-file FILE
                                                                     | --add-cc-stake-pool-id STAKE_POOL_ID
                                                                     | --add-cc-stake-verification-key STRING
                                                                     | --add-cc-stake-verification-key-file FILE
                                                                     | --add-cc-stake-key-hash HASH
                                                                     )
                                                                     --epoch NATURAL]
                                                                   --quorum RATIONAL
                                                                   [--governance-action-tx-id TXID
                                                                     --governance-action-index WORD32]
                                                                   --out-file FILE
```

**Comment by smelc** [comment]
@CarlosLopezDeLara> if I retry your [earlier example from Slack](https://input-output-rnd.slack.com/archives/C031GFLC3DF/p1695234362455249?thread_ts=1695232177.212229&cid=C031GFLC3DF), it now fails differently:

```shell
> cabal run cardano-cli -- conway governance action create-new-committee --testnet --governance-action-deposit 0 --stake-verification-key-file utxo-keys/stake1.vkey --proposal-url https://tinyurl.com/3wrwb2as --proposal-file proposal.txt --remove-cc-cold-verification-key-file cc/cc-cold.vkey --add-cc-cold-verification-key-file cc/cc-cold1.vkey --epoch 5 --quorum 
1/100 --out-file new-cc.action

cardano-cli: proposal.txt: withBinaryFile: does not exist (No such file or directory)
```

Whereas before this PR, I was reproducing your parsing failure. So I think we're good.

**Comment by smelc** [comment]
@CarlosLopezDeLara> while doing this PR, I realized we likely have the same kind of problem for `create-treasury-withdrawal`. This is visible here in this PR: https://github.com/input-output-hk/cardano-cli/blob/smelc/fix-new-committee-parser/cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs#L328 In this position, we should likely pass a prefix, because the `pAnyStakeIdentifier` parser is used twice, and so is likely ambiguous.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:105]: I think we could use `IsString` here:
```haskell
instance IsString a => IsString (Maybe a) where fromString = Just . fromString
```

this way we could just write `"remove-cc"` instead without `Just`

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakePool.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/StakePool.hs]: You can include this typo fix in this pr I think

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:328]: I agree, I think this should be included in this pr.

**Review by carbolymer — APPROVED** [review]
LGTM, but get an approval from @CarlosLopezDeLara and @Jimbo4350 or @newhoggy too

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:105]: I'm not sure, generally speaking, this instance looks surprising to me (in other words: I'd be surprised to have in my context :slightly_smiling_face:).

Also the occurrences of `Just "prefix"` are actually quite rare, most of this PR is just passing the prefix in context around.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakePool.hs
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/StakePool.hs]: Done :+1:

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:328]: I would do a separate PR, because it's a different command. I can do this other PR soon after this one is merged.

**Comment by CarlosLopezDeLara** [comment]
I can confirm that this resolved the parsing issue.

**Comment by CarlosLopezDeLara** [comment]
> @CarlosLopezDeLara> while doing this PR, I realized we likely have the same kind of problem for `create-treasury-withdrawal`. This is visible here in this PR: https://github.com/input-output-hk/cardano-cli/blob/smelc/fix-new-committee-parser/cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs#L328 In this position, we should likely pass a prefix, because the `pAnyStakeIdentifier` parser is used twice, and so is likely ambiguous.

Interestingly enough, it does not complain about this: 

```
$CARDANO_CLI conway governance action create-treasury-withdrawal \
--testnet \
--governance-action-deposit 0 \
--stake-verification-key-file "${UTXO_DIR}/stake1.vkey" \
--proposal-url https://tinyurl.com/3wrwb2as \
--proposal-file  "${ROOT}/govActionJustification.txt" \
--stake-verification-key-file "${UTXO_DIR}/stake1.vkey" \
--transfer "${WITHDRAW_AMOUNT}" \
--out-file "${UTXO_DIR}/treasury.action" 
```

**Comment by smelc** [comment]
> Interestingly enough, it does not complain about this

Weird :face_with_spiral_eyes:

**Pull request merged**
